packages/native-stack/src/views/RetainContext.tsx

```diff
@@ -0,0 +1,138 @@
+ * to return to the app.
+ */
+export const RetainContext = React.createContext<RetainContextT>({
+  retain() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  release() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  restore() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  supported: false,
+});
+
+type RetainedScenes = {
+  [key: string]: {
+    route: Route<any>;
+    descriptor: NativeStackDescriptorMap['key'];
+  };
+};
+
+export function useRetainContext(
+  state: StackNavigationState<ParamListBase>,
+  navigation: NativeStackNavigationHelpers,
+  descriptors: NativeStackDescriptorMap
+) {
+  const [retainedScenes, setRetainedScenes] = React.useState<RetainedScenes>(
+    {}
+  );
+  const latestValues = React.useRef({ state, retainedScenes, descriptors });
+  React.useEffect(() => {
+    latestValues.current = { state, retainedScenes, descriptors };
+  });
+
+  const retainContext = React.useMemo<RetainContextT>(
+    () => ({
+      retain() {
+        const { state, descriptors } = latestValues.current;
+        const route = state.routes[state.routes.length - 1];
+        setRetainedScenes((screens) => ({
+          ...screens,
+          [route.key]: {
+            route,
+            descriptor: descriptors[route.key],
+          },
+        }));
+        return route.key;
+      },
+      release(key) {
+        setRetainedScenes((screens) => {
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          const { [key]: _, ...rest } = screens;
+          return rest;
+        });
+        return latestValues.current.retainedScenes[key] != null;
+      },
+      restore(key) {
+        const { retainedScenes, state } = latestValues.current;
+
+        const route = retainedScenes[key]?.route;
+        if (!route) {
+          return false;
+        }
+
+        // Remove from retained
+        setRetainedScenes((screens) => {
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          const { [key]: _, ...rest } = screens;
+          return rest;
+        });
+
+        const index = state.routes.findIndex((r) => r.key === key);
+        const routes =
+          index !== -1 ? state.routes.slice(0, index) : state.routes.slice();
+        routes.push(route);
+        navigation.dispatch({
+          type: 'RESET',
+          payload: { ...state, index: routes.length - 1, routes },
+        });
+
+        return true;
+      },
+      supported: true,
+    }),
+    [navigation]
+  );
+
+  const routeKeys = state.routes.reduce((map, route) => {
+    map[route.key] = true;
+    return map;
+  }, {} as { [key: string]: true });
+
+  // Routes that were marked as 'retain' and have been removed from the stack,
+  // for which we still want to keep the memory allocated.
+  const hiddenRoutes: Route<any>[] = [];
+  const hiddenDescriptors: NativeStackDescriptorMap = {};
+  Object.keys(retainedScenes).forEach((key) => {
```

Comment:
I'd prefer 
```
const {
  hiddenDescriptors: descriptors,
  hiddenRoutes: routes 
} = Object.values(retainedScreen).reduce(() => { ... }, { routes: [], descriptors: {} })
```

Reply:
The code will be much simpler if there was just a `hiddenRouteKeys` array. Seems like the only place that needs it is when we set `hidden={true}`, and with such an array, it could just check `hidden={hiddenRoutes.includes(route.key)}` without changing descriptor and routes related logic.